### PR TITLE
[DSV4][BCG] Optimize the heavy memory use of C4 Indexer when BCG is enabled

### PR DESCRIPTION
**Pull request by SYChen123**
## Motivation

When enabling BCG on deepseek v4, C4 indexer leads to heavy memory pressure. The root reason is that the indexer logits with shape of [N, context_length] is padded to [N, max_position_embeddings]. For deepseek v4, model's max_position_emb is 1048576. When N=8192, the logits itself will cost 8192 * 1048576 // 4 * 4 = 8GB.
Therefore, we propose to add a new parameter named --cuda-graph-prefill-max-context and the logit is padded to the user-specified max context size instead model's maximum context.

## Modifications

1. Add --cuda-graph-prefill-max-context parameter which accepts only 1 value and currently only takes effect on deepseek v4. The c4_max_seq_len used in C4 indexer is padded to this user-specified value, rather than model's maximum context.
2. During replay, if the kv context length of any of the request exceeds the --cuda-graph-prefill-max-context, fallback to eager.

## Accuracy Tests

gsm8k test on deepseek-v4-flash-0731 is good.
image

### Memory usage reduction
Before this pr, capture will cost 43GB when --cuda-graph-max-bs-prefill=16384.
image

With this pr, the capture memory cost is reduced to 18GB. **58% memory reduction**.
image

### E2E test on deepseek-v4-flash-0731

E2E latency has 10% performance gains with BCG enabled. Without this pr, the server will OOM when chunk-prefill-size and max-captured-size are both set to 16384.

| Setup        | TTFT / TPOT (ms) | E2E (ms) | Input throughput (token/s) |
|--------------|------------------|----------|----------------------------|
| Baseline     | 1075.57 / 1.68   | 1912.69  | 278674.46                  |
| With this pr | **830.93** / 1.82    | **1737.34**  | **306275.00**                  |

Server launching and benchmark commands are shown below.
``` 
# launch server
sglang serve \
    --trust-remote-code \
    --model-path deepseek-ai/DeepSeek-V4-Flash-0731 \
    --tp 2 \
    --moe-runner-backend flashinfer_mxfp4 \
    --speculative-algorithm DSPARK \
    --chunked-prefill-size 16384 \
    --max-prefill-tokens 16384 \
    --cuda-graph-backend-prefill breakable \
  	--cuda-graph-max-bs-prefill 16384 \
	--cuda-graph-prefill-max-context 262144 \
    --enable-deepseek-v4-fp4-indexer \
    --max-running-requests 256 \
    --swa-full-tokens-ratio 0.1 \
    --mem-fraction-static 0.75 \
    --reasoning-parser deepseek-v4 \
    --tool-call-parser deepseekv4 \
    --enable-strict-thinking \
    --enable-hierarchical-cache \
    --hicache-ratio 2 \
    --enable-cache-report \
    --enable-metrics \
    --log-requests \
    --log-requests-level 1 \
    --host 192.168.153.127 \
    --port 9090 > server-0.log 2>&1 &

# benchmark
python3 -m sglang.bench_serving \
            --model deepseek-ai/DeepSeek-V4-Flash-0731 \
            --flush-cache \
            --host "${HOST}" \
            --port "${PORT}" \
            --backend sglang \
            --dataset-name generated-shared-prefix \
            --dataset-path ShareGPT_V3_unfiltered_cleaned_split.json \
            --num-prompts 512 \
            --gsp-num-groups 1 \
            --gsp-prompts-per-group 512 \
            --gsp-system-prompt-len 61440 \
            --gsp-question-len 4096 \
            --gsp-output-len 500 \
            --max-concurrency 32 \
            --warmup-requests 10 \
            > "benchmark.log" 2>&1 
    done
done
```

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
3. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
4. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
5. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :white_check_mark: [Run #34568924841](https://github.com/sgl-project/sglang/actions/runs/34568924841)
Latest PR Test (Extra): :white_check_mark: [Run #34568924736](https://github.com/sgl-project/sglang/actions/runs/34568924736)
Latest PR Test (AMD ROCm 10): :heavy_minus_sign: **No AMD PR run found for this commit**.

### CODE REVIEW

File: python/sglang/srt/model_executor/runner/prefill_cuda_graph_runner.py [hunk 1]
```diff
@@ -881,15 +947,30 @@ def _has_uncapturable_chunked_prefix(
...
+    ) -> ShapeKey:
         variant = None
         if self._capture_chunked_prefix and self._has_prefix_hit(forward_batch):
             captured_n = self._select_prefix_capture_chunks(
                 forward_batch.extend_prefix_lens_cpu
             )
             assert captured_n is not None, "prefix batch has no captured FullCG variant"
             variant = _chunked_prefix_variant(captured_n)
-        return ShapeKey(size=num_tokens, variant_label=variant)
+        if getattr(self, "capture_context_sizes", ()) and context_size is None:
```
Comment (Oasis-Git) [inline python/sglang/srt/model_executor/runner/prefill_cuda_graph_runner.py]: Do not use getattr

File: python/sglang/srt/model_executor/runner/prefill_cuda_graph_runner.py [hunk 2]
```diff
@@ -1148,6 +1231,18 @@ def can_replay_locally(
             return False
         if return_logprob and not self._uses_eager_prefill_tail():
             return False
+        if getattr(self, "capture_context_sizes", ()):
```
Comment (Oasis-Git) [inline python/sglang/srt/model_executor/runner/prefill_cuda_graph_runner.py]: Same

File: python/sglang/srt/model_executor/runner/prefill_cuda_graph_runner.py [hunk 3]
```diff
@@ -1385,32 +1485,59 @@ def _capture_one_stream(self) -> None:
...
             if get_parallel().tp_rank == 0:
                 avail_mem = get_available_gpu_memory(
                     self.model_runner.device,
                     self.model_runner.gpu_id,
                     empty_cache=False,
                 )
                 capture_range.set_description(
-                    f"Capturing num tokens ({num_tokens=} {avail_mem=:.2f} GB)"
+                    "Capturing prefill shape "
+                    f"({num_tokens=} {context_size=} {avail_mem=:.2f} GB)"
```
Comment (Oasis-Git) [inline python/sglang/srt/model_executor/runner/prefill_cuda_graph_runner.py:1526]: context size may be empty?

File: python/sglang/srt/model_executor/cuda_graph_config.py [hunk 4]
```diff
@@ -93,6 +95,12 @@ class PhaseConfig:
     bs: Optional[List[int]] = None
     # Only meaningful when backend == tc_piecewise; ignored otherwise.
     tc_compiler: str = "eager"
+    # Only meaningful for prefill body-capture backends (breakable/full):
```
Comment (Oasis-Git) [inline python/sglang/srt/model_executor/cuda_graph_config.py]: trim the comments

File: python/sglang/srt/model_executor/runner/prefill_cuda_graph_runner.py (hunk 3 above)
Comment (SYChen123) [reply python/sglang/srt/model_executor/runner/prefill_cuda_graph_runner.py:1526]: If --cuda-graph-context-bucket-prefill is not explicitly set, the context_size here printed will be None. And c4 indexer will fallback to pad seqlens to model's maximum context length, which is the same as current default behavior.

File: python/sglang/srt/model_executor/cuda_graph_config.py (hunk 4 above)
Comment (SYChen123) [reply python/sglang/srt/model_executor/cuda_graph_config.py]: Done!

File: python/sglang/srt/model_executor/runner/prefill_cuda_graph_runner.py (hunk 1 above)
Comment (SYChen123) [reply python/sglang/srt/model_executor/runner/prefill_cuda_graph_runner.py]: Done!

File: python/sglang/srt/model_executor/runner/prefill_cuda_graph_runner.py (hunk 2 above)
Comment (SYChen123) [reply python/sglang/srt/model_executor/runner/prefill_cuda_graph_runner.py]: Done.

File: python/sglang/srt/arg_groups/fields/exec_.py [hunk 5]
```diff
@@ -474,6 +475,20 @@ class ExecGraph:
         Optional[List[int]],
         "Explicit list of batch sizes to capture for the prefill cuda graph.",
     ] = None
+    cuda_graph_prefill_context_bucket: A[
```
Comment (Oasis-Git) [inline python/sglang/srt/arg_groups/fields/exec_.py]: use `cuda_graph_prefill_max_context`

File: python/sglang/srt/layers/attention/deepseek_v4_backend.py [hunk 6]
```diff
@@ -1547,9 +1558,16 @@ def prepare_forward_metadata_for_breakable_cuda_graph_replay(
         # Build graph-compatible metadata against the padded static batch. The
         # batch still carries live seq/extend lens, so the online c128 prefill
         # plan remains batch-specific without constructing a second metadata set.
+        metadata_batch = (
+            static_forward_batch if static_forward_batch is not None else forward_batch
+        )
+        max_seq_len = (
+            getattr(metadata_batch, "max_seq_len_override", None)
```
Comment (Oasis-Git) [inline python/sglang/srt/layers/attention/deepseek_v4_backend.py]: pls be careful about the use of `getattr`. refer: https://github.com/sgl-project/sglang/blob/main/.claude/rules/no-getattr-defensive.md

File: python/sglang/srt/model_executor/runner/prefill_cuda_graph_runner.py [hunk 7]
```diff
@@ -870,6 +884,73 @@ def _max_addressable_prefix_len(model_runner) -> int:
...
+            "Ignoring prefill CUDA graph max context size %d for %s; this path "
+            "does not yet preserve the fixed metadata extent.",
+            self.max_context_size,
+            execution_path,
+        )
+        self.max_context_size = None
+
+    @staticmethod
+    def _max_context_len(forward_batch: ForwardBatch) -> Optional[int]:
+        seq_lens_cpu = getattr(forward_batch, "seq_lens_cpu", None)
```
Comment (Oasis-Git) [inline python/sglang/srt/model_executor/runner/prefill_cuda_graph_runner.py]: do not use getattr

File: python/sglang/srt/model_executor/runner/prefill_cuda_graph_runner.py [hunk 8]
```diff
@@ -870,6 +884,73 @@ def _max_addressable_prefix_len(model_runner) -> int:
...
+                f"aligned size {aligned} > {max_context_len}"
+            )
+        if requested_size != aligned:
+            logger.info(
+                "Page-aligning prefill CUDA graph max context size %d -> %d "
+                "(page_size=%d).",
+                requested_size,
+                aligned,
+                page_size,
+            )
```
Comment (Oasis-Git) [inline python/sglang/srt/model_executor/runner/prefill_cuda_graph_runner.py:904]: This part should be in the server args

**Review by Oasis-Git — COMMENTED** [review]
In general I think the functionality is good. Here are some comments for code structure. Also pls update the pr description.

File: python/sglang/srt/arg_groups/fields/exec_.py (hunk 5 above)
Comment (SYChen123) [reply python/sglang/srt/arg_groups/fields/exec_.py]: Done.

File: python/sglang/srt/layers/attention/deepseek_v4_backend.py (hunk 6 above)
Comment (SYChen123) [reply python/sglang/srt/layers/attention/deepseek_v4_backend.py]: Done! The default value of max_seq_len_override is None so the getattr is removed

File: python/sglang/srt/model_executor/runner/prefill_cuda_graph_runner.py (hunk 7 above)
Comment (SYChen123) [reply python/sglang/srt/model_executor/runner/prefill_cuda_graph_runner.py]: Done

File: python/sglang/srt/model_executor/runner/prefill_cuda_graph_runner.py (hunk 8 above)
Comment (SYChen123) [reply python/sglang/srt/model_executor/runner/prefill_cuda_graph_runner.py:904]: Done. The _resolve_max_context_size is moved to server args.

**Review by Oasis-Git — APPROVED** [review]

**Comment by Oasis-Git** [comment]
Let's run ci

**Comment by Oasis-Git** [comment]
extra pass: https://github.com/sgl-project/sglang/actions/runs/34568924736

**Comment by Oasis-Git** [comment]
base pass: https://github.com/sgl-project/sglang/actions/runs/34568924841
